test/AutoDiff/Sema/derivative_attr_type_checking.swift

```diff
@@ -442,24 +463,129 @@ extension Struct {
+  func vjpSubscriptGenericGetter<T: Differentiable>(x: T) -> (value: T, pullback: (T.TangentVector) -> TangentVector)   {
+    return (x, { _ in .zero })
+  }
+
+  // expected-error @+2 {{a derivative already exists for '_'}}
+  // expected-note @-6 {{other attribute declared here}}
   @derivative(of: subscript(_:), wrt: self)
-  func vjpSubscriptGeneric<T: Differentiable>(x: T) -> (value: T, pullback: (T.TangentVector) -> TangentVector) {
+  func vjpSubscriptGeneric<T: Differentiable>(x: T) -> (value: T, pullback: (T.TangentVector) -> TangentVector)   {
     return (x, { _ in .zero })
   }
+
+
+  @derivative(of: subscript.set)
+  mutating func vjpSubscriptSetter(_ newValue: Float) -> (
+    value: (), pullback: (inout TangentVector) -> Float
+  ) {
+    fatalError()
+  }
+
+  @derivative(of: subscript().set)
+  mutating func jvpSubscriptSetter(_ newValue: Float) -> (
+    value: (), differential: (inout TangentVector, Float) -> ()
+  ) {
+    fatalError()
+  }
+
+  @derivative(of: subscript(float:).set)
+  mutating func vjpSubscriptLabeledSetter(float: Float, newValue: Float) -> (
+    value: (), pullback: (inout TangentVector) -> (Float, Float)
+  ) {
+    fatalError()
+  }
+
+  @derivative(of: subscript(float:).set)
+  mutating func jvpSubscriptLabeledSetter(float: Float, _ newValue: Float) -> (
+    value: (), differential: (inout TangentVector, Float, Float) -> Void
+  ) {
+    fatalError()
+  }
+
+  // Error: original subscript has no setter.
+  // expected-error @+1 {{'subscript(_:)' does not have a 'set' accessor}}
+  @derivative(of: subscript(_:).set, wrt: self)
+  mutating func vjpSubscriptGeneric_NoSetter<T: Differentiable>(x: T) -> (
+    value: T, pullback: (T.TangentVector) -> TangentVector
+  ) {
+    return (x, { _ in .zero })
+  }
+}
+
+extension Class {
+  subscript() -> Float {
+    get { 1 }
+    // expected-note @+1 {{'subscript()' declared here}}
+    set {}
+  }
+}
+extension Class where T: Differentiable {
+  @derivative(of: subscript.get)
+  func vjpSubscriptGetter() -> (value: Float, pullback: (Float) -> TangentVector) {
+    return (1, { _ in .zero })
+  }
+
+  // expected-error @+2 {{a derivative already exists for '_'}}
+  // expected-note @-6 {{other attribute declared here}}
+  @derivative(of: subscript)
+  func vjpSubscript() -> (value: Float, pullback: (Float) -> TangentVector) {
+    return (1, { _ in .zero })
+  }
+
+  // FIXME(SR-13096): Enable derivative registration for class property/subscript setters.
+  // This requires changing derivative type calculation rules for functions with
+  // class-typed parameters. We need to assume that all functions taking
+  // class-typed operands may mutate those operands.
+  // expected-error @+1 {{cannot yet register derivative for class property or subscript setters}}
+  @derivative(of: subscript.set)
+  func vjpSubscriptSetter(_ newValue: Float) -> (
+    value: (), pullback: (inout TangentVector) -> Float
+  ) {
+    fatalError()
+  }
```

Comment:
In addition to `@derivative` attribute type-checking tests, we should add some end-to-end differentiation tests to verify that accessor derivative registration fully works.

Reply:
Done in f1f60345b47d729d34afcd3ba52f20cb8b149f68.